pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/ConverterUtil.java

```diff
@@ -0,0 +1,385 @@
+
+  public static LocalTime toLocalTime(Object o) {
+    if (o instanceof LocalTime) {
+      return (LocalTime) o;
+    }
+    if (o instanceof Time) {
+      return ((Time) o).toLocalTime();
+    }
+    if (o instanceof byte[]) {
+      return Schema.LOCAL_TIME.decode((byte[]) o);
+    }
+    if (o instanceof CharSequence) {
+      return LocalTime.parse((CharSequence) o);
+    }
+    try {
+      return LocalTime.ofInstant(toInstant(o), ZoneOffset.UTC);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException(
+          "Cannot convert type " + o.getClass() + " to LocalTime", e);
+    }
+  }
+
+  public static LocalDate toLocalDate(Object o) {
+    if (o instanceof LocalDate) {
+      return (LocalDate) o;
+    }
+    if (o instanceof LocalDateTime) {
+      return ((LocalDateTime) o).toLocalDate();
+    }
+    if (o instanceof byte[]) {
+      return Schema.LOCAL_DATE.decode((byte[]) o);
+    }
+    if (o instanceof CharSequence) {
+      return LocalDate.parse((CharSequence) o);
+    }
+    try {
+      Instant instant = toInstant(o);
+      return LocalDate.ofInstant(instant, ZoneOffset.UTC);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException(
+          "Cannot convert type " + o.getClass() + " to LocalDate", e);
+    }
+  }
+
+  public static LocalDateTime toLocalDateTime(Object o) {
+    if (o instanceof LocalDateTime) {
+      return (LocalDateTime) o;
+    }
+    if (o instanceof LocalDate) {
+      return ((LocalDate) o).atStartOfDay();
+    }
+    if (o instanceof byte[]) {
+      return Schema.LOCAL_DATE_TIME.decode((byte[]) o);
+    }
+    if (o instanceof CharSequence) {
+      return LocalDateTime.parse((CharSequence) o);
+    }
+    try {
+      Instant instant = toInstant(o);
+      return LocalDateTime.ofInstant(instant, ZoneOffset.UTC);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException(
+          "Cannot convert type " + o.getClass() + " to LocalDateTime", e);
+    }
+  }
+
+  public static Instant toInstant(Object o) {
+    if (o instanceof Instant) {
+      return (Instant) o;
+    }
+    if (o instanceof LocalDate) {
+      return ((LocalDate) o).atStartOfDay().toInstant(ZoneOffset.UTC);
+    }
+    if (o instanceof LocalDateTime) {
+      return ((LocalDateTime) o).toInstant(ZoneOffset.UTC);
+    }
+    if (o instanceof OffsetDateTime) {
+      return ((OffsetDateTime) o).toInstant();
+    }
+    if (o instanceof Date && !(o instanceof Time)) {
+      return ((Date) o).toInstant();
+    }
+    if (o instanceof Long) {
+      return Instant.ofEpochMilli(((Number) o).longValue());
+    }
+    if (o instanceof Double) {
+      long seconds = (long) ((double) o / 1000);
+      long nanos = Math.round(((double) o - seconds * 1000) * 1_000_000);
+      return Instant.ofEpochSecond(seconds, nanos);
+    }
+    if (o instanceof CharSequence) {
+      return DateTimeFormatter.ISO_DATE_TIME.parse((CharSequence) o, Instant::from);
```

Comment:
I just checked checked how OffsetDateTime works and it actually uses [ISO_OFFSET_DATE_TIME](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_OFFSET_DATE_TIME) . There are nuances between the two that I don't fully understand but we can agree on to use it consistently between the cast and compute steps.